swarm_transport.go

```diff
@@ -0,0 +1,61 @@
+package swarm
+
+import (
+	"fmt"
+	"strings"
+
+	transport "github.com/libp2p/go-libp2p-transport"
+	ma "github.com/multiformats/go-multiaddr"
+)
+
+// TransportForAddr retrieves the appropriate transport for the given multiaddr.
+func (s *Swarm) TransportForAddr(a ma.Multiaddr) transport.Transport {
+	protocols := a.Protocols()
+	if len(protocols) == 0 {
+		return nil
+	}
+
+	s.transports.RLock()
+	defer s.transports.RUnlock()
+	for _, p := range protocols {
+		transport, ok := s.transports.m[p.Code]
+		if !ok {
+			continue
+		}
+		if transport.Proxy() {
+			return transport
+		}
+	}
+	p := protocols[len(protocols)-1]
```

Comment:
Is there any possibility that a multiaddress can specify no protocols? If yes, this may panic.

Reply:
(yes, it would be very weird that such multiaddr would pop-up here but still)

Reply:
There's a check for that at the top of the function (although it shouldn't happen).